File: DevExpress-Demo/CS/GridDemo.Wpf/Modules/VirtualSources/InfiniteScrolling2ViewModel.cs
using DevExpress.Mvvm;
using DevExpress.Mvvm.DataAnnotations;
using DevExpress.Mvvm.Xpf;
using DevExpress.Xpf.Data;
using System;
using System.ComponentModel;
using System.Linq;
using System.Threading.Tasks;

namespace GridDemo.VirtualSources {
    public class InfiniteScrolling2ViewModel : ViewModelBase {
        [Command]
        public void FetchIssues(FetchRowsAsyncArgs args) {
            args.Result = FetchRowsAsync(args);
        }
        static async Task<FetchRowsResult> FetchRowsAsync(FetchRowsAsyncArgs args) {
            #region !
            IssueSortOrder sortOrder = IssueSortOrder.Default;
            if(args.SortOrder.Length > 0) {
                var sort = args.SortOrder.Single();
                if(sort.PropertyName == "Created") {
                    if(sort.Direction != ListSortDirection.Descending)
                        throw new InvalidOperationException();
                    sortOrder = IssueSortOrder.CreatedDescending;
                }
                if(sort.PropertyName == "Votes") {
                    sortOrder = sort.Direction == ListSortDirection.Ascending
                        ? IssueSortOrder.VotesAscending
                        : IssueSortOrder.VotesDescending;
                }
            }
            #endregion

            #region Service request
            int take = args.Take ?? 50;
            IssueFilter filter = null;
            var issues = await IssuesService.GetIssuesAsync(
                skip: args.Skip,
                take: take,
                sortOrder: sortOrder,
                filter: filter);

            return new FetchRowsResult(rows: issues, hasMoreRows: issues.Length == take);
            #endregion
        }
    }
}

File: DevExpress-Demo/VB/GridDemo.Wpf/Modules/VirtualSources/InfiniteScrolling2ViewModel.vb
Imports DevExpress.Mvvm
Imports DevExpress.Mvvm.DataAnnotations
Imports DevExpress.Mvvm.Xpf
Imports DevExpress.Xpf.Data
Imports System
Imports System.ComponentModel
Imports System.Linq
Imports System.Threading.Tasks

Namespace GridDemo.VirtualSources

    Public Class InfiniteScrolling2ViewModel
        Inherits ViewModelBase

        <Command>
        Public Sub FetchIssues(ByVal args As FetchRowsAsyncArgs)
            args.Result = FetchRowsAsync(args)
        End Sub

        Private Shared Async Function FetchRowsAsync(ByVal args As FetchRowsAsyncArgs) As Task(Of FetchRowsResult)
#Region "!"
            Dim sortOrder As IssueSortOrder = IssueSortOrder.Default
            If args.SortOrder.Length > 0 Then
                Dim sort = args.SortOrder.[Single]()
                If Equals(sort.PropertyName, "Created") Then
                    If sort.Direction <> ListSortDirection.Descending Then Throw New InvalidOperationException()
                    sortOrder = IssueSortOrder.CreatedDescending
                End If

                If Equals(sort.PropertyName, "Votes") Then
                    sortOrder = If(sort.Direction = ListSortDirection.Ascending, IssueSortOrder.VotesAscending, IssueSortOrder.VotesDescending)
                End If
            End If

#End Region
#Region "Service request"
            Dim take As Integer = If(args.Take, 50)
            Dim filter As IssueFilter = Nothing
            Dim issues = Await GetIssuesAsync(skip:=args.Skip, take:=take, sortOrder:=sortOrder, filter:=filter)
            Return New FetchRowsResult(rows:=issues, hasMoreRows:=issues.Length = take)
#End Region
        End Function
    End Class
End Namespace
